File: frontend/src/types/strategic.ts
export type ProjectType = 'project' | 'program' | 'plan' | 'digital_transform'
export type ProjectStatus = 'planning' | 'active' | 'on_hold' | 'completed' | 'cancelled'
export type PriorityLevel = 'low' | 'medium' | 'high' | 'critical'
export type BudgetStatus = 'draft' | 'approved' | 'active' | 'closed' | 'over_budget'
export type FundingType =
  | 'ngan_sach_tinh'
  | 'ngan_sach_xa'
  | 'trung_uong'
  | 'von_dau_tu'
  | 'xa_hoi_hoa'
  | 'tai_tro'

export interface UserMin {
  id: number
  username: string
  full_name: string | null
}

export interface DeptMin {
  id: number
  name: string
  short_name: string | null
}

export interface StaffMin {
  id: number
  full_name: string
  position: string | null
  employee_code: string | null
  department_id: number | null
}

// ── Strategic Project ────────────────────────────────────────────────────────

export interface StrategicProject {
  id: number
  project_code: string | null
  project_name: string
  project_type: ProjectType
  program_id: number | null
  nghi_quyet_id: number | null
  source_document_id: number | null
  source_document: { id: number; title: string; doc_number: string | null } | null
  muc_tieu_id: number | null
  description: string | null
  start_date: string | null
  end_date: string | null
  project_status: ProjectStatus
  priority_level: PriorityLevel
  progress_percent: number
  responsible_department_id: number | null
  coordinating_departments: DeptMin[]
  project_manager_id: number | null
  project_manager_staff_id: number | null
  responsible_department: DeptMin | null
  project_manager: UserMin | null
  project_manager_staff: StaffMin | null
  creator: UserMin | null
  created_at: string
  updated_at: string
}

export interface StrategicProjectCreate {
  project_code?: string
  project_name: string
  project_type?: ProjectType
  program_id?: number | null
  nghi_quyet_id?: number
  source_document_id?: number | null
  muc_tieu_id?: number
  description?: string
  start_date?: string
  end_date?: string
  project_status?: ProjectStatus
  priority_level?: PriorityLevel
  progress_percent?: number
  responsible_department_id?: number
  coordinating_department_ids?: number[]
  project_manager_id?: number
  project_manager_staff_id?: number | null
}

export interface StrategicProjectUpdate extends Partial<StrategicProjectCreate> {}

export interface StrategicProjectList {
  total: number
  items: StrategicProject[]
}

// ── Budget Plan ──────────────────────────────────────────────────────────────

export interface BudgetPlan {
  id: number
  budget_code: string | null
  project_id: number
  fiscal_year: number
  total_budget: number
  allocated_budget: number
  spent_budget: number
  remaining_budget: number
  budget_status: BudgetStatus
  note: string | null
  creator: UserMin | null
  created_at: string
  updated_at: string
}

export interface BudgetPlanCreate {
  budget_code?: string
  project_id: number
  fiscal_year: number
  total_budget?: number
  allocated_budget?: number
  spent_budget?: number
  budget_status?: BudgetStatus
  note?: string
}

export interface BudgetPlanUpdate extends Partial<Omit<BudgetPlanCreate, 'project_id'>> {}

export interface BudgetPlanList {
  total: number
  items: BudgetPlan[]
}

// ── Funding Source ───────────────────────────────────────────────────────────

export interface FundingSource {
  id: number
  budget_plan_id: number
  funding_source_name: string
  funding_type: FundingType
  funding_amount: number
  funding_year: number | null
  note: string | null
  created_at: string
}

export interface FundingSourceCreate {
  budget_plan_id: number
  funding_source_name: string
  funding_type?: FundingType
  funding_amount?: number
  funding_year?: number
  note?: string
}

export interface FundingSourceList {
  total: number
  items: FundingSource[]
}

// ── Disbursement ─────────────────────────────────────────────────────────────

export interface Disbursement {
  id: number
  disbursement_code: string | null
  budget_plan_id: number
  disbursement_date: string
  disbursement_amount: number
  evidence_file: string | null
  note: string | null
  creator: UserMin | null
  created_at: string
}

export interface DisbursementCreate {
  disbursement_code?: string
  budget_plan_id: number
  disbursement_date: string
  disbursement_amount: number
  evidence_file?: string
  note?: string
}

export interface DisbursementUpdate extends Partial<Omit<DisbursementCreate, 'budget_plan_id'>> {}

export interface DisbursementList {
  total: number
  items: Disbursement[]
}

// ── Dashboard Stats ──────────────────────────────────────────────────────────

export interface StrategicDashboardStats {
  total_projects: number
  active_projects: number
  completed_projects: number
  on_hold_projects: number
  planning_projects: number
  total_budget: number
  total_allocated: number
  total_spent: number
  total_remaining: number
  disbursement_rate: number
  avg_progress: number
  overdue_projects: number
  by_status: Record<string, number>
  by_type: Record<string, number>
  by_priority: Record<string, number>
  top_slow_projects: Array<{
    id: number
    project_name: string
    progress_percent: number
    end_date: string
    priority_level: PriorityLevel
  }>
}

// ── Labels ───────────────────────────────────────────────────────────────────

export const PROJECT_TYPE_LABELS: Record<ProjectType, string> = {
  project: 'Dự án',
  program: 'Chương trình',
  plan: 'Kế hoạch',
  digital_transform: 'Chuyển đổi số',
}

export const PROJECT_STATUS_LABELS: Record<ProjectStatus, string> = {
  planning: 'Lập kế hoạch',
  active: 'Đang thực hiện',
  on_hold: 'Tạm dừng',
  completed: 'Hoàn thành',
  cancelled: 'Đã hủy',
}

export const PRIORITY_LABELS: Record<PriorityLevel, string> = {
  low: 'Thấp',
  medium: 'Trung bình',
  high: 'Cao',
  critical: 'Cấp thiết',
}

export const BUDGET_STATUS_LABELS: Record<BudgetStatus, string> = {
  draft: 'Dự thảo',
  approved: 'Đã duyệt',
  active: 'Hoạt động',
  closed: 'Đã đóng',
  over_budget: 'Vượt ngân sách',
}

export const FUNDING_TYPE_LABELS: Record<FundingType, string> = {
  ngan_sach_tinh: 'Ngân sách tỉnh',
  ngan_sach_xa: 'Ngân sách xã',
  trung_uong: 'Trung ương',
  von_dau_tu: 'Vốn đầu tư',
  xa_hoi_hoa: 'Xã hội hóa',
  tai_tro: 'Tài trợ',
}

File: backend/app/schemas/strategic.py
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, field_validator


class UserMin(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    username: str
    full_name: str | None = None


class DeptMin(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    name: str
    short_name: str | None = None


class StaffMin(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    full_name: str
    position: str | None = None
    employee_code: str | None = None
    department_id: int | None = None


# ─── StrategicProject ────────────────────────────────────────────────────────

class StrategicProjectCreate(BaseModel):
    project_code: str | None = None
    project_name: str
    project_type: str = "project"
    program_id: int | None = None
    nghi_quyet_id: int | None = None
    source_document_id: int | None = None
    muc_tieu_id: int | None = None
    description: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    project_status: str = "planning"
    priority_level: str = "medium"
    progress_percent: int = 0
    responsible_department_id: int | None = None
    coordinating_department_ids: list[int] | None = None
    project_manager_id: int | None = None
    project_manager_staff_id: int | None = None


class StrategicProjectUpdate(BaseModel):
    project_code: str | None = None
    project_name: str | None = None
    project_type: str | None = None
    program_id: int | None = None
    nghi_quyet_id: int | None = None
    source_document_id: int | None = None
    muc_tieu_id: int | None = None
    description: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    project_status: str | None = None
    priority_level: str | None = None
    progress_percent: int | None = None
    responsible_department_id: int | None = None
    coordinating_department_ids: list[int] | None = None
    project_manager_id: int | None = None
    project_manager_staff_id: int | None = None


class DocMin(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    title: str
    doc_number: str | None = None


class StrategicProjectOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    project_code: str | None = None
    project_name: str
    project_type: str
    program_id: int | None = None
    nghi_quyet_id: int | None = None
    source_document_id: int | None = None
    source_document: DocMin | None = None
    muc_tieu_id: int | None = None
    description: str | None = None
    start_date: date | None = None
    end_date: date | None = None
    project_status: str
    priority_level: str
    progress_percent: int
    responsible_department_id: int | None = None
    coordinating_departments: list[DeptMin] = []
    project_manager_id: int | None = None
    project_manager_staff_id: int | None = None
    responsible_department: DeptMin | None = None
    project_manager: UserMin | None = None
    project_manager_staff: StaffMin | None = None
    creator: UserMin | None = None
    created_at: datetime
    updated_at: datetime


class StrategicProjectList(BaseModel):
    model_config = {"from_attributes": True}
    total: int
    items: list[StrategicProjectOut]


# ─── BudgetPlan ──────────────────────────────────────────────────────────────

class BudgetPlanCreate(BaseModel):
    budget_code: str | None = None
    project_id: int
    fiscal_year: int
    total_budget: float = 0.0
    allocated_budget: float = 0.0
    spent_budget: float = 0.0
    budget_status: str = "draft"
    note: str | None = None


class BudgetPlanUpdate(BaseModel):
    budget_code: str | None = None
    fiscal_year: int | None = None
    total_budget: float | None = None
    allocated_budget: float | None = None
    spent_budget: float | None = None
    budget_status: str | None = None
    note: str | None = None


class BudgetPlanOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    budget_code: str | None = None
    project_id: int
    fiscal_year: int
    total_budget: float
    allocated_budget: float
    spent_budget: float
    remaining_budget: float
    budget_status: str
    note: str | None = None
    creator: UserMin | None = None
    created_at: datetime
    updated_at: datetime


class BudgetPlanList(BaseModel):
    model_config = {"from_attributes": True}
    total: int
    items: list[BudgetPlanOut]


# ─── FundingSource ───────────────────────────────────────────────────────────

class FundingSourceCreate(BaseModel):
    budget_plan_id: int
    funding_source_name: str
    funding_type: str = "ngan_sach_xa"
    funding_amount: float = 0.0
    funding_year: int | None = None
    note: str | None = None


class FundingSourceUpdate(BaseModel):
    funding_source_name: str | None = None
    funding_type: str | None = None
    funding_amount: float | None = None
    funding_year: int | None = None
    note: str | None = None


class FundingSourceOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    budget_plan_id: int
    funding_source_name: str
    funding_type: str
    funding_amount: float
    funding_year: int | None = None
    note: str | None = None
    created_at: datetime


class FundingSourceList(BaseModel):
    model_config = {"from_attributes": True}
    total: int
    items: list[FundingSourceOut]


# ─── Disbursement ────────────────────────────────────────────────────────────

class DisbursementCreate(BaseModel):
    disbursement_code: str | None = None
    budget_plan_id: int
    disbursement_date: date
    disbursement_amount: float
    evidence_file: str | None = None
    note: str | None = None


class DisbursementUpdate(BaseModel):
    disbursement_code: str | None = None
    disbursement_date: date | None = None
    disbursement_amount: float | None = None
    evidence_file: str | None = None
    note: str | None = None


class DisbursementOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    disbursement_code: str | None = None
    budget_plan_id: int
    disbursement_date: date
    disbursement_amount: float
    evidence_file: str | None = None
    note: str | None = None
    creator: UserMin | None = None
    created_at: datetime


class DisbursementList(BaseModel):
    model_config = {"from_attributes": True}
    total: int
    items: list[DisbursementOut]


# ─── ProjectTaskLink ─────────────────────────────────────────────────────────

class ProjectTaskLinkCreate(BaseModel):
    project_id: int
    task_id: int


class ProjectTaskLinkOut(BaseModel):
    model_config = {"from_attributes": True}
    id: int
    project_id: int
    task_id: int
    created_at: datetime


# ─── Dashboard / Stats ───────────────────────────────────────────────────────

class StrategicDashboardStats(BaseModel):
    total_projects: int
    active_projects: int
    completed_projects: int
    on_hold_projects: int
    planning_projects: int
    total_budget: float
    total_allocated: float
    total_spent: float
    total_remaining: float
    disbursement_rate: float        # spent / total_budget * 100
    avg_progress: float
    overdue_projects: int
    by_status: dict[str, int]
    by_type: dict[str, int]
    by_priority: dict[str, int]
    top_slow_projects: list[dict]   # projects with low progress near deadline
